Orange/data/table.py

```diff
@@ -33,146 +28,176 @@ def get_sample_datasets_dir():
+        """
+        return self._to_numpy(X=True)
 
     @property
     def Y(self):
-        if self._Y.shape[1] == 1:
-            return self._Y[:, 0]
-        return self._Y
+        """
+        Return a read-only numpy array/matrix of Y.
+        If there is only one column, returns a one-dimensional array,
+            otherwise, returns a two-dimensional matrix whose columns
+            are in the same order as returned by Table.columns_Y.
+        """
+        res = self._to_numpy(Y=True)
+        # TODO: comment on whether this separation is okay please
+        return res.flatten() if len(self.columns_Y) else res
+
+    @property
+    def metas(self):
+        """
+        Return a read-only numpy matrix of meta attributes.
+        Always a two-dimensional matrix, even for a single column.
+        The columns are in the same order as returned by Table.columns_metas.
+        """
+        return self._to_numpy(meta=True)
+
+    @property
+    def weights(self):
+        # TODO: do we switch to .weights and deprecate .W or keep .W?
+        """Get the weights as a numpy array."""
+        return self[Table._WEIGHTS_COLUMN].values
 
-    @Y.setter
-    def Y(self, value):
-        if len(value.shape) == 1:
-            value = value[:, None]
-        self._Y = value
+    @property
+    def W(self):
+        return self.weights
+
+    def set_role(self, column_names, column_roles):
+        """
+        Sets the role a column (or multiple columns) takes in this table.
+        """
+        if isinstance(column_names, str):
+            column_names = [column_names]
+        if isinstance(column_roles, str):
+            column_roles = [column_roles]
+        for n, rstr in zip(column_names, column_roles):
+            r = Role.from_string(rstr)
+            self._columns_X.discard(n)
+            self._columns_Y.discard(n)
+            self._columns_meta.discard(n)
+            if r == Role.x:
+                self._columns_X.add(n)
+            elif r == Role.y:
+                self._columns_Y.add(n)
+            elif r == Role.meta:
+                self._columns_meta.add(n)
+            else:
+                raise ValueError("{} is not a valid role name".format(rstr))
+
+    def set_weights(self, weight):
+        """
+        Set the weights for the instances in this table.
+        If a number, weights to set to that value.
+        If a string, weights are set to whatever the column with that name's values are,
+            but only if those values are all numbers and are not NA/NaN.
+        If a sequence of (non-NA/NaN) numbers, set those values as the sequence.
+        """
+        if np.isreal(weight):
+            self[Table._WEIGHTS_COLUMN] = weight
+        elif isinstance(weight, str):
+            if weight not in self.columns:
+                raise ValueError("{} is not a column.".format(weight))
+            if not self[weight].applymap(np.isreal).all() or self[weight].isnull().any():
```

Comment:
Why median and not mode? The requirement for weights to be non-nan is reasonable. Never mind.